rfcs/2021-01-08-5843-encdoing-decoding-format-vector.md

```diff
@@ -0,0 +1,95 @@
+# RFC 5843 - 2021-01-08 - Encoding/Decoding for Vector to Vector Communication
+
+There has been an ongoing discussion of what format Vector-to-Vector communications should take. For some examples of issues related to the subject check out #5809, #5124, or #5341. This RFC is intended to propose a format for standardization of our vector-to-vector communications.
+
+## Scope
+
+- This RFC should cover the encoding format we intend to leverage
+- It should also cover any alterations that might need to be made to the expected data format
+
+## Motivation
+
+The question posed by @binarylogic in #5843 is a good definition for the motivation of this change: "How do we transmit data from one Vector instance to another while being mindful of the public contract we create?" Clearly sending metrics from one Vector instance to another should preserve all metric data upon ingestion downstream but there are still concerns around exposing a proprietary format to users in the context of these operations. This discussion continues to come up and we really need a canonincal and definitive decision around what shape this takes so that when we expand our components our contract is explicitly codified.
+
+## Internal Proposal
+
+From a high level my proposal is pretty simple. I'll dive into the rationale and context lower in the doc but lets just tee that up, shall we?
+
+I propose that if we're not searching explicitly for better perf around these messages and we're not explicitly searching for alternate tooling because we've decided that we wholly can't deal with the tooling overhead of protobufs we instead continue to use them with prost. I would also posit that if the concerns around exposing our private data format are bothersome enough, we should consider following an open spec for the data as a primary example: [opentelemetry](https://opentelemetry.io/).
+
+Should performance be a heavier concern for this standardization than I have assumed for the writing of this doc, my propsal would be to consider a highly performant alternative to protobufs like Flatbuffers and still consider following an open spec for the data. The `Rationale` section below will cover my thoughts on this more thoroughly.
```

Comment:
It is not clear if flatbuffers retain their performance advantage when the features that allow backwards compatibility are enabled. That is, while naive structs can be accessed much faster than protobufs, does that speed advantage remain if tables are used in their stead to allow for the changing of fields. I don't think we can take others' word on its performance unless they have tested with tables that require some decoding effort.

It's also worth considering that we have to consider both the encoding and decoding of the data equally, since vector bears both costs. While flatbuffers appears to be much faster to decode (subject to the above), but both prost and bincode are faster to encode, and bincode has a combined speed that tops flatbuffers overall.

All of this is to say we can't really trust the numbers below to be anything close to representative for our case.

Reply:
Do I understand correctly that NoProto is essentially the idea of dumping the memory behind the data structures one's already using to the wire? In that sense, it would be very similar to FlatBuffers.

UPD: nope, it's a different thing - a format of it's own. We should consider the wild idea of just dumping the memory to the wire though - it might be doable, with some internal changes, and could be extremely efficient.

Reply:
I'm not sure just duping the memory to the wire is feasible, or at least not without some rewriting. Consider the case of strings. They are stored as a structure containing a capacity and length and a *pointer* to a block of UTF-8 bytes. We have to stream both the structure and the bytes, but they are in separate locations. Should the pointer be rewritten to contain an offset into the written block where the bytes can be found or do we omit it and assume the bytes are immediately follow the string length? Do we need to store the capacity?

Even in something simple like a string, there are data layout decisions that go beyond just dumping out the memory. This also doesn't deal with cross-version compatibility (how to dump data from one version and reload with another) or more complex structures. We need a well designed/defined protocol with an efficient implementation, so we had best use something existing.

Reply:
Oh yes, for sure not without rewrites. Thinking about it, it doesn't seem like it would be practical as a real-world implementation. However, dumping memory might be just an edge case to compare the perf to, as we can't be any quicker than that in theory.